qoala/sim/scheduler.py

```diff
@@ -283,53 +362,206 @@ def wait(self, delta_time: float) -> Generator[EventExpression, None, None]:
+            ev_expr = self.await_signal(self._cpu_scheduler, SIGNAL_TASK_COMPLETED)
+            ev_expr = ev_expr | self.await_signal(
+                self._qpu_scheduler, SIGNAL_TASK_COMPLETED
+            )
+            yield ev_expr
+
+            now = ns.sim_time()
+
+            # Gets the pid of the last finished task at the current time,
+            # if there is no task that is finished at the current time, it returns -1
+            self._last_cpu_task_pid = self.cpu_scheduler.get_last_finished_task_pid_at(
+                now
+            )
+            # If there is a task that is finished at the current time, assign the next
+            if self._last_cpu_task_pid != -1:
+                self.schedule_next_for(self._last_cpu_task_pid)
+
+            self._last_qpu_task_pid = self.qpu_scheduler.get_last_finished_task_pid_at(
+                now
+            )
+            # If there is a task that is finished at the current time, assign the next, but we do not need to assign
+            # again if it is the same pid as the one that came from CPU
+            if (
+                self._last_qpu_task_pid != -1
+                and self._last_qpu_task_pid != self._last_cpu_task_pid
+            ):
+                self.schedule_next_for(self._last_qpu_task_pid)
+
+    def schedule_next_for(self, pid: int) -> None:
+        """
+        Schedule the tasks of the next block for program instance with given pid
+        by assigning respective tasks to CPU and QPU schedulers and sends a message
+        to schedulers for informing them about the newly assigned tasks.
+
+        :param pid: program instance id
+        :return: None
+        """
+        new_cpu_tasks, new_qpu_tasks = self.find_next_tasks_for(pid)
+
+        # If there are new tasks, send a message to schedulers
+        # Note that find_next_tasks_for() returns None if there are no new tasks for that processor
+        if new_cpu_tasks:
+            self._cpu_scheduler.add_tasks(new_cpu_tasks)
+            self._comp.send_cpu_scheduler_message(Message(-1, -1, "New Task"))
+        if new_qpu_tasks:
+            self._qpu_scheduler.add_tasks(new_qpu_tasks)
+            self._comp.send_qpu_scheduler_message(Message(-1, -1, "New Task"))
+
+    def schedule_all(self) -> None:
+        """
+        Schedules the tasks of the next block for each available program instance in the memory manager
+        by assigning respective tasks to CPU and QPU schedulers and sends a message
+        to schedulers for informing them about the newly assigned tasks.
+
+        This method is responsible for scheduling tasks for each available program instance in the memory manager.
+        A program instance is considered available if it meets two conditions:
+        1. It is not finished.
+        2. It does not have any dependencies on an unfinished program instance
+        which is the case if the batch of program instances are submitted to run linearly.
```

Comment:
just to avoid ambiguity I would rewrite it a little (right now it's not clear if the "which is the case" applies to "having dependencies" or "not having dependencies")
```suggestion
        (it can have such dependencies if the batch of program instances are submitted to run linearly).
```